uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/AbsencePeriod.php

```diff
@@ -308,28 +310,127 @@ public function getNumberOfWorkingDays()
+    if (strtotime($endDate) > strtotime($this->end_date)) {
       $endDate = $this->end_date;
     }
 
-    $periodToWork = new CRM_HRLeaveAndAbsences_BAO_AbsencePeriod();
+    $periodToWork             = new CRM_HRLeaveAndAbsences_BAO_AbsencePeriod();
     $periodToWork->start_date = $startDate;
-    $periodToWork->end_date = $endDate;
+    $periodToWork->end_date   = $endDate;
 
     return $periodToWork->getNumberOfWorkingDays();
   }
+
+  /**
+   * Returns the Absence Period previous to this one. That is, the Absence
+   * Period where weight is equal to this Period weight - 1.
+   *
+   * @return null|CRM_HRLeaveAndAbsences_BAO_AbsencePeriod - The previous Absence Period or null if there's none
+   */
+  public function getPreviousPeriod()
+  {
+    $previousPeriod = new CRM_HRLeaveAndAbsences_BAO_AbsencePeriod();
+    $previousPeriod->weight = $this->weight - 1;
+    $previousPeriod->find(true);
+    if($previousPeriod->id) {
+      return $previousPeriod;
+    }
+
+    return null;
+  }
+
+  /**
+   * Calculates the expiration date for the given AbsenceType within this period.
+   *
+   * If the AbsenceType has an expiration duration (that is,
+   * carry_forward_expiration_duration and carry_forward_expiration_unit are not
+   * empty), its value is used to calculate the expiration date starting from
+   * the period start date.
+   *
+   * @param \CRM_HRLeaveAndAbsences_BAO_AbsenceType $absenceType
+   *
+   * @return null|string A date in Y-m-d or null if it can not be calculated
+   */
+  public function getExpirationDateForAbsenceType(AbsenceType $absenceType)
+  {
+    if(!$this->hasValidDates()) {
+      throw new UnexpectedValueException(
+        'You can only calculate the expiration date for an AbsenceType from an AbsencePeriod with start and end dates'
+      );
+    }
+
+    $expirationDate = null;
+
+    if($absenceType->hasExpirationDuration()) {
+      $expirationDate = $this->getExpirationDurationDate($absenceType);
+    }
+
+    return $expirationDate;
+  }
+
+  /**
+   * Returns the expiration date calculated based on the AbsenceType expiration
+   * duration.
+   *
+   * Example: If the expiration duration is 5 days and this AbsencePeriod
+   * start_date is 2016-01-01, the expiration date will be 2016-01-06.
+   *
+   * @param \CRM_HRLeaveAndAbsences_BAO_AbsenceType $absenceType
+   *
+   * @return null|string A date in Y-m-d or null if it can not be calculated
+   */
+  private function getExpirationDurationDate(AbsenceType $absenceType)
+  {
+    if(!$absenceType->allow_carry_forward || !$absenceType->hasExpirationDuration()) {
+      return null;
+    }
+
+    switch($absenceType->carry_forward_expiration_unit) {
+      case AbsenceType::EXPIRATION_UNIT_DAYS:
+        $unit = 'D';
+        break;
+      case AbsenceType::EXPIRATION_UNIT_MONTHS:
+        $unit = 'M';
+        break;
+      case AbsenceType::EXPIRATION_UNIT_YEARS:
+        $unit = 'Y';
+        break;
+      default:
+        $unit = null;
```

Comment:
Maybe I am mistaken but I can see that $unit can only have one of 4 values which are ( 'D', 'M' , 'Y' or null)  .. and in case it was null , you are returning null .